examples/www-image/main/jpg-render.c

```diff
@@ -0,0 +1,518 @@
+    uint8_t g = *(bitmap_ptr++);
+    uint8_t b = *(bitmap_ptr++);
+
+    // Calculate weighted grayscale
+    //uint32_t val = ((r * 30 + g * 59 + b * 11) / 100); // original formula
+    uint32_t val = (r*38 + g*75 + b*15) >> 7; // @vroland recommended formula
+
+    int xx = rect->left + i % w;
+    if (xx < 0 || xx >= image_width) {
+      continue;
+    }
+    int yy = rect->top + i / w;
+    if (yy < 0 || yy >= image_width) {
+      continue;
+    }
+    
+    /* Optimization note: If we manage to apply here the epd_draw_pixel directly
+       then it will be no need to keep a huge raw buffer (But will loose dither) */
+    decoded_image[yy * image_width + xx] = gamme_curve[val];
+  }
+
+  return 1;
+}
+
+//====================================================================================
+//   This function opens source_buf Jpeg image file and primes the decoder
+//====================================================================================
+int drawBufJpeg(uint8_t *source_buf, int xpos, int ypos) {
+  rc = jd_prepare(&jd, feed_buffer, tjpgd_work, sizeof(tjpgd_work), &source_buf);
+  if (rc != JDR_OK) {    
+    ESP_LOGE(TAG, "JPG jd_prepare error: %s", jd_errors[rc]);
+    return ESP_FAIL;
+  }
+
+  uint32_t decode_start = esp_timer_get_time();
+
+  // Last parameter scales        v 1 will reduce the image
+  rc = jd_decomp(&jd, tjd_output, 0);
+  if (rc != JDR_OK) {
+    ESP_LOGE(TAG, "JPG jd_decomp error: %s", jd_errors[rc]);
+    return ESP_FAIL;
+  }
+
+  time_decomp = (esp_timer_get_time() - decode_start)/1000;
+
+  ESP_LOGI("JPG", "width: %d height: %d\n", jd.width, jd.height);
+  ESP_LOGI("decode", "%d ms . image decompression", time_decomp);
+
+  // Render the image onto the screen at given coordinates
+  jpegRender(xpos, ypos, jd.width, jd.height);
+
+  return 1;
+}
+
+// Handles Htpp events and is in charge of buffering source_buf (jpg compressed image)
+esp_err_t _http_event_handler(esp_http_client_event_t *evt)
+{
+    uint8_t output_buffer[HTTP_RECEIVE_BUFFER_SIZE]; // Buffer to store HTTP response
+
+    switch (evt->event_id)
+    {
+    case HTTP_EVENT_ERROR:
+        ESP_LOGE(TAG, "HTTP_EVENT_ERROR");
+        break;
+    case HTTP_EVENT_ON_CONNECTED:
+        ESP_LOGI(TAG, "HTTP_EVENT_ON_CONNECTED");
+        break;
+    case HTTP_EVENT_HEADER_SENT:
+        ESP_LOGI(TAG, "HTTP_EVENT_HEADER_SENT");
+        break;
+    case HTTP_EVENT_ON_HEADER:
+        #if DEBUG_VERBOSE
+          ESP_LOGI(TAG, "HTTP_EVENT_ON_HEADER, key=%s, value=%s", evt->header_key, evt->header_value);
+        #endif
+        break;
+    case HTTP_EVENT_ON_DATA:
+        ++countDataEventCalls;
+        #if DEBUG_VERBOSE
+          if (countDataEventCalls%10==0) {
+            ESP_LOGI(TAG, "%d len:%d\n", countDataEventCalls, evt->data_len); 
+          }
+        #endif
+        dataLenTotal += evt->data_len;
+        
+        // Copy the response into the buffer
+        memcpy(output_buffer, evt->data, evt->data_len);
```

Comment:
Add a check that evt->data_len does not overflow output_buffer.

Reply:
Done, check my comments in the conversation below, I think it’s not really necessary since the buffer_size http configuration should already cover this.